sycl/source/detail/queue_impl.hpp

```diff
@@ -1022,6 +1027,35 @@ class queue_impl {
     }
   } MDefaultGraphDeps, MExtGraphDeps;
 
+  // implement check-lock-check pattern to not lock empty MData
+  template <typename DataType> class CheckLockCheck {
+    DataType MData;
+    std::atomic_bool MDataPresent = false;
+    mutable std::mutex MDataMtx;
+
+  public:
+    template <typename F> void put(F &&func) {
+      std::lock_guard<std::mutex> Lock(MDataMtx);
+      MDataPresent.store(true, std::memory_order_release);
+      func(MData);
+    }
+    template <typename F> void get(F &&func) {
```

Comment:
I think modifications on `get` might be unexpected. Would `pop` work as a name?

Reply:
Done. Then, it might be reasonable to change `put` to `pop`.

Reply:
Did you mean `push`?

Reply:
Sorry, typo. Yes, push/pop instead of put/get.